tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ConcreteFunction.java

```diff
@@ -0,0 +1,291 @@
+   *       Tensor<TFloat32> x = TFloat32.scalarOf(2.0f)) {
+   *     assertEquals(4.0f, function.call(x).expect(TFloat32.DTYPE).data().getFloat());
+   *   }
+   *   // Graph g is still valid at this point
+   * }
+   * }</pre>
+   *
+   * @param signature signature of the function to create
+   * @param graph a valid and initialized graph
+   * @return a new function
+   */
+  public static ConcreteFunction create(Signature signature, Graph graph) {
+    return new ConcreteFunction(signature, graph, new Session(graph), Ownership.SESSION_ONLY);
+  }
+
+  /**
+   * Create a function from a signature and a valid graph session.
+   *
+   * <p/>The function will not own the session nor its graph, meaning that their lifetime
+   * can extend beyond the scope of the function. Therefore the function does not need to be
+   * closed after its usage. For example:
+   *
+   * <pre>{@code
+   * try (Graph g = new Graph()) {
+   *   Placeholder<TFloat32> input = tf.placeholder(TFloat32.DTYPE);
+   *   Add<TFloat32> output = tf.math.add(input, tf.constant(2.0f));
+   *   Signature signature = Signature.builder().input("x", input).output("y", output).build();
+   *
+   *   try (Session s = new Session(g)) {
+   *     // Auto-closing the function just as an example but this is not required since it has
+   *     // no effect
+   *     try (ConcreteFunction f = ConcreteFunction.create(signature, s);
+   *         Tensor<TFloat32> t = TFloat32.scalarOf(2.0f)) {
+   *       assertEquals(4.0f, function.call(x).expect(TFloat32.DTYPE).data().getFloat());
+   *     }
+   *     // Session s is still valid at this point
+   *   }
+   *   // Graph g is still valid at this point
+   * }
+   * }</pre>
+   *
+   * @param signature signature of the function to create
+   * @param graph a valid session to an initialized graph
+   * @return a new function
+   */
+  public static ConcreteFunction create(Signature signature, Session session) {
+    return new ConcreteFunction(signature, session.graph(), session, Ownership.NONE);
+  }
+
+  /**
+   * Returns the signature of this function
+   */
+  public Signature signature() {
+    return signature;
+  }
+
+  /**
+   * Invokes a function.
+   *
+   * <p>Caller is responsible for closing all Tensors.
+   *
+   * @param tensor input tensor
+   * @return output tensor
+   */
+  public Map<String, Tensor<?>> call(Map<String, Tensor<?>> arguments)
+      throws IllegalArgumentException {
+
+    final SignatureDef signatureDef = signature.asSignatureDef();
+    final Session.Runner runner = session.runner();
+
+    signatureDef.getInputsMap().forEach((argName, t) -> {
+      Tensor<?> tensor = arguments.get(argName);
+      if (tensor == null) {
+        throw new IllegalArgumentException(String.format("Missing argument [%s]", argName));
+      }
+      runner.feed(t.getName(), tensor);
+    });
+
+    Map<String, TensorInfo> outputToNode = signatureDef.getOutputsMap();
```

Comment:
The ordering of signature's output map need not be consistent.

We depend on the ordering out output returned in the call  `List<Tensor<?>> resultTensors = runner.run()`. Which depends on the order we pass in at `runner.fetch(t.getName())`, that happens to be the iteration order of `signatureDef.getOutputsMap()`.

See references
[runHelper](https://sourcegraph.com/github.com/tensorflow/java@09c136ea671153452217138629e30cf83dc169f6/-/blob/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java#L323)

See creation of `outputOpIndices` based on order of `outputs.add(output)`, which gets passed into `Session.run(..)`

Session.run uses `outputOpIndices` as the order of output, which is implemented in `resolveOutputs` [see] (https://sourcegraph.com/github.com/tensorflow/java@09c136ea671153452217138629e30cf83dc169f6/-/blob/tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Graph.java?utm_source=share#L606:7)

Reply:
So it will be consistent within an execution of `call` but could change across different VM executions if the `signatureDef.getOutputsMap()` is backed by a regular HashMap rather than a LinkedHashMap?

I ask because we have a tendency to iterate the value set of maps in demo code, and I don't want users to think those will be consistent if they aren't. Usually it's because there is only a single output, but if we're not maintaining ordering then people might be confused  if they try to apply what we do in the demos/tests to larger problems.

Reply:
The order we fetch the tensors does not really matter, as we don't return the outputs as a list but as a map, forcing the user to retrieve them by name instead of by ordinal position, like `Session.Run` does. And since this is not demo code neither, I think it is safe enough to leave it as is.

Reply:
Is this really correct? One line does `outputToNode.values().forEach(t -> runner.fetch(t.getName()));` and the other `for (String nodeName: outputToNode.keySet()) {` - is there any guarantee of consistent ordering between values() and keySet()!? 

The code very much depends on this and there's no test to confirm (ConcreteFunctionTest.java only uses the simpler call()-method).

Reply:
Protobuf uses `LinkedHashMap` internally, so yeah it's fine.  Wouldn't be a bad idea to convert it to one manually though.

Reply:
Maybe use `.values()` in both to raise fewer red flags? Though relying on a `Map` producing the same values() order is perhaps also questionable?

Reply:
Not if it's a `LinkedHashMap`, that's the point of the class, it provides stable iteration over a Map with minimal overhead.